google/cloud/internal/rest_completion_queue.cc

```diff
@@ -0,0 +1,80 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/rest_completion_queue.h"
+#include "google/cloud/options.h"
+#include <mutex>
+#include <thread>
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+void RestCompletionQueue::Shutdown() {
+  std::lock_guard<std::mutex> lock(mu_);
+  shutdown_ = true;
+}
+
+RestCompletionQueue::QueueStatus RestCompletionQueue::GetNext(
+    void** tag, bool* ok, std::chrono::system_clock::time_point deadline) {
+  auto check_queue = [&]() mutable {
+    std::lock_guard<std::mutex> lock(mu_);
+    if (shutdown_) {
+      *tag = nullptr;
+      *ok = false;
+      return QueueStatus::kShutdown;
+    }
+
+    if (!pending_tags_.empty()) {
+      auto* front = std::move(pending_tags_.front());
+      pending_tags_.pop_front();
+      *tag = front;
+      *ok = true;
+      return QueueStatus::kGotEvent;
+    }
+    return QueueStatus::kTimeout;
+  };
+
+  // This is a naive implementation of using the deadline to return if no tags
+  // are present.
+  auto initial_check = check_queue();
+  if (initial_check != QueueStatus::kTimeout) return initial_check;
+  std::this_thread::sleep_until(deadline);
```

Comment:
`sleep_until()` has an odor, the prime one being: don't you want this to unblock early if someone does an `AddTag()` in the meantime?

It will also make the API harder to test.